dss/index/es/document.py

```diff
@@ -229,35 +248,40 @@ def get_shape_descriptor(self) -> typing.Optional[str]:
-                               f"as is, without sanitization. This may prevent subsequent, valid files from being "
-                               f"indexed correctly.")
+
+        def shape_rejection(file_name, schema):
+            # Enforce the prerequisites that make the mapping to shape descriptors bijective. This will enable
+            # us to parse shape descriptors should we need to in the future. Dots have to be avoided because
+            # they are used as separators. A number (the schema version) is used to terminate each file's
+            # entry in the shape descriptor, allowing us to distinguish between the normal form of an entry and
+            # the compressed form that is used when schema and file name are the same.
+            reject('.' in file_name, f"A metadata file name must not contain '.' characters: {file_name}")
+            reject(file_name.isdecimal(), f"A metadata file name must contain at least one non-digit: {file_name}")
+            reject('.' in schema.type, f"A schema name must not contain '.' characters: {schema.type}")
+            reject(schema.type.isdecimal(), f"A schema name must contain at least one non-digit: {schema.type}")
+            assert '.' not in schema.version, f"A schema version must not contain '.' characters: {schema.version}"
+            assert schema.version.isdecimal(), f"A schema version must consist of digits only: {schema.version}"
+
+        schemas_by_file: typing.Set[typing.Tuple[str, SchemaInfo]] = set()
+        for file_name, file_list in self.files.items():
+            for file_content in file_list:
+                schema = SchemaInfo.from_json(file_content)
+                if schema is not None:
+                    if file_name.endswith('_json'):
+                        file_name = file_name[:-5]
+                    shape_rejection(file_name, schema)
+                    schemas_by_file.add((file_name, schema))
+                else:
+                    logger.warning(f"Unable to obtain JSON schema info from file '{file_name}'. The file will be "
+                                   f"indexed as is, without sanitization. This may prevent subsequent, valid files "
+                                   f"from being indexed correctly.")
         if schemas_by_file:
-            same_version = 1 == len(set(schema.version for schema in schemas_by_file.values()))
-            same_schema_and_file_name = all(file_name == schema.type for file_name, schema in schemas_by_file.items())
+            same_version = 1 == len(set(schema.version for _, schema in schemas_by_file))
+            same_schema_and_file_name = all(file_name == schema.type for file_name, schema in schemas_by_file)
             if same_version and same_schema_and_file_name:
-                return 'v' + next(iter(schemas_by_file.values())).version
+                return 'v' + schemas_by_file.pop()[1].version
```

Comment:
I find it confusing to follow when you have more than one return path. Also reducing the number of returns in this case will allow us to add the hashing of the `shape_description` into `get_shape_descriptor` without adding duplicate code. Here is what I suggest:
```python
        if schemas_by_file:
            ...
            if same_version and same_schema_and_file_name:
                shape_descriptor = 'v' + schemas_by_file.pop()[1].version
            else:
                ...     
                shape_descriptor = 'v.' + '.'.join(entry(*schema) for schema in schemas)
            return hashlib.sha1(str(shape_descriptor).encode("utf-8")).hexdigest()
        else:
            return None  # No files with schema references were found
```
This will eliminate the need for L216:L217 and keep `shape_descriptor` generation code out of `_prepare_index`